### PR TITLE
docs: add how-to-write-fuzz-tests.md

### LINKED ISSUE #3451: Add a doc "how to write fuzz test"
Do we have the doc on "how to write fuzz test"? @WenyXu

_Originally posted by @MichaelScofield in https://github.com/GreptimeTeam/greptimedb/issues/3441#issuecomment-1982313627_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
close #3451 

## What's changed and what's your intention?

Add a doc to introduce how to write fuzz tests in GreptimeDB.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
+This document introduce how to write fuzz tests in GreptimeDB.
+
```
Comment (killme2008) [inline docs/how-to/how-to-write-fuzz-tests.md:2]: Let's add some descriptions about fuzz tests, what it is, and why we need them.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+All fuzz test-related resources are located in the `/tests-fuzz` directory.
+- Each target(located in the `/tests-fuzz/targets`) stands for an independent fuzz test case.
+- Fuzz test fundamental components are located in the `/tests-fuzz/src`.
+
+Figure 1 illustrates that it utilizes a Rng (Random Number Generator) to generate the IR (Intermediate Representation), then employs a DialectTranslator to produce specified dialects for different protocols.
```
Comment (killme2008) [inline docs/how-to/how-to-write-fuzz-tests.md:7]: Produce specified dialects for different protocols., then what? Looks like missing something.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+[[bin]]
+name = "{target name}"
+path = "targets/{target}.rs"
+test = false
+bench = false
+doc = false
+...
+```
+
+3. Define the `FuzzInput`
```
Comment (killme2008) [inline docs/how-to/how-to-write-fuzz-tests.md:48]: Where to put this code?

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+impl Arbitrary<'_> for FuzzInput {
+    fn arbitrary(u: &mut Unstructured<'_>) -> arbitrary::Result<Self> {
+        let seed = u.int_in_range(u64::MIN..=u64::MAX)?;
+        Ok(FuzzInput { seed })
+    }
+}
+```
+
+4. Your first fuzz test target
```
Comment (killme2008) [inline docs/how-to/how-to-write-fuzz-tests.md:64]: The same question as above.

**Review by killme2008 — COMMENTED** [review]
Great work! However, I think it needs more information and detailed step-by-step instructions to help the developers.

**Comment by tisonkun** [comment]
@PragmaTwice I found you left a comment at https://github.com/GreptimeTeam/greptimedb/issues/3174. Perhaps you'd be interested in this patch that describes how we write fuzz tests also. Welcome to drop your comments and share the experience you did fuzz tests in Apache Kvrocks and more.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
+This document introduce how to write fuzz tests in GreptimeDB.
+
+## What is the fuzz test
+Fuzz tests are tools that leverage deterministic random generation to assist in finding bugs. The goal of fuzz tests is to identify inputs generated by the fuzzer that cause system panics, crashes, or unexpected behaviors to occur.
+
```
Comment (killme2008) [inline docs/how-to/how-to-write-fuzz-tests.md:5]: Please adds some reference docs about the fuzz test framework or develop guide.

**Review by killme2008 — APPROVED** [review]
LGTM Almostly.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
+This document introduce how to write fuzz tests in GreptimeDB.
+
+## What is the fuzz test
```
Comment (etolbakov) [inline docs/how-to/how-to-write-fuzz-tests.md:3]: I would suggest "What is a fuzz test"

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+All fuzz test-related resources are located in the `/tests-fuzz` directory.
+There are two type of resources:
+
+### Fundamental components 
+They are located in the `/tests-fuzz/src` directory. The fundamental components define how to generate SQLs (including dialects for different protocols) and validate execution results (e.g., column attribute validation), etc.
+
+### Test targets
+They are located in the `/tests-fuzz/targets` directory, with each file representing an independent fuzz test case. The target utilizes fundamental components to generate SQLs, sends the generated SQLs via specified protocol, and validates the results of SQL execution.
+
+Figure 1 illustrates that the fundamental components of the fuzz test provide the ability to generate random SQLs. It utilizes a Random Number Generator (Rng) to generate the Intermediate Representation (IR), then employs a DialectTranslator to produce specified dialects for different protocols. Finally, the fuzz tests send the generated SQL via the specified protocol and verify that the execution results meet expectations.
```
Comment (etolbakov) [inline docs/how-to/how-to-write-fuzz-tests.md:21]: `that` - is not needed : `Figure 1 illustrates the fundamental`

**Review by etolbakov — COMMENTED** [review]
just a couple of nits I found while I was reading the document

**Comment by killme2008** [comment]
@etolbakov @tisonkun PTAL

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
+This document introduce how to write fuzz tests in GreptimeDB.
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:1]:
```suggestion
# How to write fuzz tests?

This document introduces how to write fuzz tests in GreptimeDB.
```

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+## What is a fuzz test
+Fuzz test is tool that leverage deterministic random generation to assist in finding bugs. The goal of fuzz tests is to identify inputs generated by the fuzzer that cause system panics, crashes, or unexpected behaviors to occur. And we are using the [cargo-fuzz](https://github.com/rust-fuzz/cargo-fuzz) to run our fuzz test targets. 
+
+## Why we need them
+- Find bugs leverage random generation
+- Integrate with other tests(e.g., e2e)
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:8]:
```suggestion
- Find bugs by leveraging random generation
- Integrate with other tests (e.g., e2e)
```

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+## Resources
+All fuzz test-related resources are located in the `/tests-fuzz` directory.
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:11]:
```suggestion
All fuzz test-related resources are located in the `tests-fuzz` directory.
```

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+2. Register the fuzz test target in the `/tests-fuzz/Cargo.toml` file.
+
+```toml
+[[bin]]
+name = "{fuzz target name}"
+path = "targets/<fuzz target name>.rs"
+test = false
+bench = false
+doc = false
+
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:64]:
```suggestion
```

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  Fuzz Test
+
+```
+(Figure1: Overview of fuzz tests)
+
+For more details about fuzz targets and fundamental components, please refer to this [tracking issue](https://github.com/GreptimeTeam/greptimedb/issues/3174).
+
+## How to add a fuzz test target
+
+1. Create an empty rust source file under the `/tests-fuzz/targets/<fuzz target name>.rs` directory.
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:53]:
```suggestion
1. Create an empty rust source file under the `/tests-fuzz/targets/<fuzz-target>.rs` directory.
```

May use a shorter and no whitespace placeholder. Ditto below.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    merge_two_word_map_fn, random_capitalize_map, uppercase_and_keyword_backtick_map,
+    MappedGenerator, WordGenerator,
+};
+use tests_fuzz::generator::create_expr::CreateTableExprGeneratorBuilder;
+use tests_fuzz::generator::Generator;
+use tests_fuzz::ir::CreateTableExpr;
+use tests_fuzz::translator::mysql::create_expr::CreateTableExprTranslator;
+use tests_fuzz::translator::DslTranslator;
+use tests_fuzz::utils::{init_greptime_connections, Connections};
+...
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:103]:
```suggestion
```

Comment (tisonkun) [reply docs/how-to/how-to-write-fuzz-tests.md:103]: Keep it compilable.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+fuzz_target!(|input: FuzzInput| {
+    common_telemetry::init_default_ut_logging();
+    common_runtime::block_on_write(async {
+        let Connections { mysql } = init_greptime_connections().await;
+            let mut rng = ChaChaRng::seed_from_u64(input.seed);
+            let columns = rng.gen_range(2..30);
+            ...
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:111]:
```suggestion
// ...?
```

Use comment to describe anything that should be added here.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+                )))
+                .columns(columns)
+                .engine("mito")
+                .if_not_exists(if_not_exists)
+                .build()
+                .unwrap();
+            let ir = create_table_generator.generate(&mut rng);
+            let translator = CreateTableExprTranslator;
+            let sql = translator.translate(&expr).unwrap();
+            ...
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:125]:
```suggestion

```

ditto

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+            mysql.execute(&sql).await
+    })
+});
+```
+
+5. Run your fuzz test target
+
+Please follow the instructions in this [document](../../tests-fuzz/README.md).
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:133]:
```suggestion
Please follow the instructions in this [document](/tests-fuzz/README.md).
```

Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md]: We can duplicate a happy path here, like:

```
cargo fuzz run fuzz_create_table --fuzz-dir tests-fuzz
```

and add the ref to test-fuzz's README for other more complicated cases.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 17]
```diff
@@ -100,15 +102,14 @@ use tests_fuzz::ir::CreateTableExpr;
...
-...
 
 fuzz_target!(|input: FuzzInput| {
     common_telemetry::init_default_ut_logging();
     common_runtime::block_on_write(async {
         let Connections { mysql } = init_greptime_connections().await;
             let mut rng = ChaChaRng::seed_from_u64(input.seed);
             let columns = rng.gen_range(2..30);
-            ...
+// ...?
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:112]: @WenyXu This suggestion doesn't for applied directly. Please read the original comment.

File: docs/how-to/how-to-write-fuzz-tests.md (hunk 12 above)
Comment (tisonkun) [reply docs/how-to/how-to-write-fuzz-tests.md]: @WenyXu this should be applied Ditto below.

**Comment by WenyXu** [comment]
@tisonkun PTAL

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+                .columns(columns)
+                .engine("mito")
+                .if_not_exists(if_not_exists)
+                .build()
+                .unwrap();
+            let ir = create_table_generator.generate(&mut rng);
+            let translator = CreateTableExprTranslator;
+            let sql = translator.translate(&expr).unwrap();
+            mysql.execute(&sql).await
+            // other logical
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:126]:
```suggestion
```

No need. Only necessary if it's something MUST DO in the framework but should be completed by users. It's common that users can add any customized logics.

File: docs/how-to/how-to-write-fuzz-tests.md [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+    })
+});
+```
+
+5. Run your fuzz test target
+
+```bash
+    cargo fuzz run <fuzz-test> --fuzz-dir tests-fuzz
+```
```
Comment (tisonkun) [inline docs/how-to/how-to-write-fuzz-tests.md:135]:
```suggestion
5. Run your fuzz test target:

```bash
cargo fuzz run <fuzz-target> --fuzz-dir tests-fuzz
```
```

The placeholder is the same as `fuzz-target`?

Comment (WenyXu) [reply docs/how-to/how-to-write-fuzz-tests.md:135]: Good catch!

**Review by tisonkun — APPROVED** [review]
Thank you! Let's land it first and improve continuously :D

**Pull request merged**